includes/class-sensei-course-enrolment.php

```diff
@@ -0,0 +1,335 @@
+<?php
+/**
+ * File containing the class Sensei_Course_Enrolment.
+ *
+ * @package sensei
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Handles course enrolment logic.
+ */
+class Sensei_Course_Enrolment {
+	const META_PREFIX_ENROLMENT_RESULTS = 'course-enrolment-';
+
+	/**
+	 * Courses instances.
+	 *
+	 * @var static[]
+	 */
+	private static $instances = [];
+
+	/**
+	 * All course enrolment providers.
+	 *
+	 * @var Sensei_Course_Enrolment_Provider_Interface[]
+	 */
+	private static $enrolment_providers;
+
+	/**
+	 * Enrolment providers handling this particular course.
+	 *
+	 * @var Sensei_Course_Enrolment_Provider_Interface[]
+	 */
+	private $course_enrolment_providers;
+
+	/**
+	 * Course enrolment providers version hash.
+	 *
+	 * @var string
+	 */
+	private $course_enrolment_providers_version;
+
+	/**
+	 * Course ID for this enrolment object.
+	 *
+	 * @var int
+	 */
+	private $course_id;
+
+	/**
+	 * Sensei_Course_Enrolment constructor.
+	 *
+	 * @param int $course_id Course ID to handle checks for.
+	 */
+	private function __construct( $course_id ) {
+		$this->course_id = $course_id;
+	}
+
+	/**
+	 * Get instance for a particular course.
+	 *
+	 * @param int $course_id Course ID to handle checks for.
+	 *
+	 * @return self
+	 */
+	public static function get_course_instance( $course_id ) {
+		if ( ! isset( self::$instances[ $course_id ] ) ) {
+			self::$instances[ $course_id ] = new static( $course_id );
+		}
+
+		return self::$instances[ $course_id ];
+	}
+
+	/**
+	 * Gets the course ID for this enrolment object.
+	 *
+	 * @return int
+	 */
+	public function get_course_id() {
+		return $this->course_id;
+	}
+
+	/**
+	 * Check if a user is enroled in a course.
+	 *
+	 * @param int  $user_id       User ID.
+	 * @param bool $force_recheck Force a recalculation with all providers.
+	 * @return bool
+	 */
+	public function is_enroled( $user_id, $force_recheck = false ) {
+		try {
+			if ( ! $force_recheck ) {
```

Comment:
I switched the signature in 222ed12f413a29682b02a5d6281e552d7b7fe8ef to be positive instead of negative... hopefully that helps readability. I wasn't sure how to remove the second nested `if` without doing inline assignment or nesting all the negative code (when cache wasn't available). Definitely open to more suggestions though!

Reply:
What I meant was something like the following code. But I think your solution is also good. Good approach! =)

```php
if ( $force_check ) {
	$enrolment_check_results = $this->query_enrolment_providers( $user_id );
	$is_enrolled             = $enrolment_check_results->is_enrolment_provided();
	
	// @todo Method will always be bool when we add manual enrolment provider. Remove this and the `get_legacy_enrolment_status` temporary method.
	if ( ! is_bool( $is_enrolled ) ) {
		$is_enrolled = $this->get_legacy_enrolment_status( $user_id );
	}
	
	$this->save_enrolment( $user_id, $is_enrolled );
	return $is_enrolled;
}

$enrolment_check_results = $this->get_enrolment_check_results( $user_id );
if (
	$enrolment_check_results
	&& $enrolment_check_results->get_version_hash() === $this->get_course_enrolment_providers_version()
) {
	return $this->has_stored_enrolment( $user_id );
}
```